File: demos/java/DataStructuresDemo.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class DataStructuresDemo {
    record Task(String name, int priority) implements Comparable<Task> {
        public int compareTo(Task other) {
            return Integer.compare(priority, other.priority);
        }
    }

    record Node<T>(T value, List<Node<T>> children) {}

    public static void main(String[] args) {
        List<String> queue = new ArrayList<>(List.of("parse", "validate", "render"));
        Map<String, Integer> counts = new LinkedHashMap<>();
        queue.forEach(step -> counts.merge(step.substring(0, 1), 1, Integer::sum));

        PriorityQueue<Task> tasks = new PriorityQueue<>();
        tasks.add(new Task("docs", 2));
        tasks.add(new Task("tests", 1));

        Node<String> tree = new Node<>("root", List.of(
            new Node<>("left", List.of()), new Node<>("right", List.of())
        ));

        System.out.println("list=" + queue);
        System.out.println("counts=" + counts);
        System.out.println("next=" + tasks.remove().name());
        System.out.println("tree-children=" + tree.children().size());
    }
}

File: demos/php/data_structures_demo.php
<?php

final readonly class Node {
    /** @param list<Node> $children */
    public function __construct(public string $value, public array $children = []) {}
}

$orderedMap = ['Ada' => 95, 'Lin' => 91];
$queue = new SplQueue();
$queue->enqueue('parse');
$queue->enqueue('render');
$heap = new SplPriorityQueue();
$heap->setExtractFlags(SplPriorityQueue::EXTR_BOTH);
$heap->insert('docs', -2);
$heap->insert('tests', -1);
$tree = new Node('root', [new Node('left'), new Node('right')]);

echo 'map=' . implode(',', array_map(fn ($k, $v) => "$k:$v", array_keys($orderedMap), $orderedMap)) . "\n";
echo 'queue-first=' . $queue->dequeue() . "\n";
echo 'next=' . $heap->extract()['data'] . "\n";
echo 'tree-children=' . count($tree->children) . "\n";
